Implementing catalog update acl commands

anusuyar opened the pull request:
```
Implementing following commands

vcd catalog acl add <catalog-name> [subject:name:right]*
vcd catalog acl remove <catalog-name> [subject:name]*
vcd catalog acl share <catalog-name> --access-level ReadOnly
vcd catalog acl unshare <catalog-name>
vcd catalog acl list
vcd catalog acl info

C:\sources\github\service_ecosystem\vcd-cli (master)
(vcd-cli) λ vcd catalog acl add Test2 'org:AnuOrg1:ReadOnly'  'user:user:FullControl'
subject_name         subject_type    access_level
-------------------  --------------  --------------
AnuOrg (org_in_use)  org             None
AnuOrg1              org             ReadOnly
user                 user            FullControl

C:\sources\github\service_ecosystem\vcd-cli (master) 
(vcd-cli) λ vcd catalog acl remove Test2 'org:AnuOrg1'
subject_name         subject_type    access_level
-------------------  --------------  --------------
AnuOrg (org_in_use)  org             None
user                 user            FullControl

C:\sources\github\service_ecosystem\vcd-cli (master) 
(vcd-cli) λ vcd catalog acl remove Test2 --all
Do you want to remove all acl from the catalog 'Test2' [y/N]: y
subject_name         subject_type    access_level
-------------------  --------------  --------------
AnuOrg (org_in_use)  org             None


C:\sources\github\service_ecosystem\vcd-cli (master) 
(vcd-cli) λ vcd catalog acl share Test2
subject_name         subject_type    access_level
-------------------  --------------  --------------
AnuOrg (org_in_use)  org             ReadOnly

C:\sources\github\service_ecosystem\vcd-cli (master) 
(vcd-cli) λ vcd catalog acl unshare Test2
subject_name         subject_type    access_level
-------------------  --------------  --------------
AnuOrg (org_in_use)  org             None

C:\sources\github\service_ecosystem\vcd-cli (master) 
(vcd-cli) λ vcd catalog acl add Test2 'org:AnuOrg1:ReadOnly'  'user:user:FullControl'
subject_name         subject_type    access_level
-------------------  --------------  --------------
AnuOrg (org_in_use)  org             None
AnuOrg1              org             ReadOnly
user                 user            FullControl


C:\sources\github\service_ecosystem\vcd-cli (master) 
(vcd-cli) λ vcd catalog acl list Test2
subject_name         subject_type    access_level
-------------------  --------------  --------------
AnuOrg (org_in_use)  org             None
anuorg1              org             ReadOnly
user                 user            FullControl

C:\sources\github\service_ecosystem\vcd-cli (master) 
(vcd-cli) λ vcd catalog acl info Test2
property                        value
------------------------------  -----------------------------------------------------------------------------------------------------
access_settings_1_access_level  ReadOnly
access_settings_1_subject_href  https://bos1-vcd-sp-static-199-204.eng.vmware.com/api/admin/org/1992f0d8-cb4a-434e-b150-a87e96ba3cd0
access_settings_1_subject_name  anuorg1
access_settings_1_subject_type  application/vnd.vmware.admin.organization+xml
access_settings_2_access_level  FullControl
access_settings_2_subject_href  https://bos1-vcd-sp-static-199-204.eng.vmware.com/api/admin/user/d69620d1-1be4-41bc-8db3-ef1006dac028
access_settings_2_subject_name  user
access_settings_2_subject_type  application/vnd.vmware.admin.user+xml
is_shared_to_everyone           False

```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vmware/pyvcloud/163)
<!-- Reviewable:end -->

rocknes commented:
> vcd catalog acl list
> vcd catalog acl info

should be 

vcd catalog acl list [catalog name]
vcd catalog acl info [catalog name]

The semantics of info and list command for vcd catalog acl seems to be different from what it has been for catalog or vapp.

i.e. 'list' lists all available catalogs/vapps and 'info' picks up one of those items and provides more info on it.
The info command here seems to be working on a group of acl rules. I think we should discuss this design since it seems to be deviating from the standard.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-:-L1yw6xJQU_Xt2F96FOU:btda42c)*
<!-- Sent from Reviewable.io -->

rocknes commented:
Reviewed 3 of 4 files at r1.
Review status: 3 of 4 files reviewed at latest revision, 13 unresolved discussions.

---

*[pyvcloud/vcd/acl.py, line 29 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L2-00kebISHl1KYO7t0:-L2-00kebISHl1KYO7t1:batcsg7) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L29)):*
> ```Python
>         :param client: (pyvcloud.vcd.client): The client.
>         :param resource: (lxml.objectify.ObjectifiedElement): XML
>             representation of the entity in which acl belongs to.
> ```

nit : XML representation of the 'parent' entity whose Access Control List this object operates on.

---

*[pyvcloud/vcd/acl.py, line 32 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L2-0uDIHbE5LJJ-5ROM:-L2-0uDJwngu8DZPz8DV:bohmjbn) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L32)):*
> ```Python
>         """
>         self.client = client
>         self.resource = resource
> ```

I wouldn't call this resource, rather parent_resource. resource would be the XML we get when we do a GET on resource['href']/accesscontrol

---

*[pyvcloud/vcd/acl.py, line 35 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L2-2b2k066hIPpU_mU9:-L2-2b2k066hIPpU_mUA:bopblyq) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L35)):*
> ```Python
> 
>     def add_access_settings(self, access_settings_list=None):
>         """Add acl to resource.
> ```

Append a *new* access control list to the existing ACL of the parent resource

---

*[pyvcloud/vcd/acl.py, line 53 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L2-3IHgU7T_GwXzriGn:-L2-3IHgU7T_GwXzriGo:bnz0uyx) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L53)):*
> ```Python
> 
>         # if access_settings_list is None, nothing to add.
>         if access_settings_list is not None:
> ```

This check should be the first thing we do in this function, else we will be making a ReST call to vCD to get control_access_param for no reason.
Infact the control_access_param is the resource in this context! So if we cache that, we don't need to make the ReST call.

---

*[pyvcloud/vcd/acl.py, line 60 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L2-351JKgDkqet6nixf:-L2-351JKgDkqet6nixg:b-pqp73v) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L60)):*
> ```Python
>                     AccessSettings
> 
>                 # discard the AccessSettings fro control_access_params as we
> ```

typo : fro ?

---

*[pyvcloud/vcd/acl.py, line 66 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L2-5kbeiZgoeioTpNmR:-L2-5kbeiZgoeioTpNmS:b-7thjjx) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L66)):*
> ```Python
>             # remove common access_setting between access_settings_list
>             # and old_access_settings_params
>             for access_setting in list(access_settings_list):
> ```

Is this conversion to list necessary? access_settings_list should be a list , right?

---

*[pyvcloud/vcd/acl.py, line 92 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L2-672rpHKg4keObceJ:-L2-672rpHKg4keObceK:b51k76t) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L92)):*
> ```Python
>                 EntityType.CONTROL_ACCESS_PARAMS.value, control_access_params)
> 
>         return control_access_params
> ```

We should change the if condition so that the cache-ed control_access_params is returned if access_settings_list if None, that will let us to indent the whole block of code above to left.

---

*[pyvcloud/vcd/client.py, line 171 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L1zXm0LVgdhf6kGP4gy:-L1zXm0LVgdhf6kGP4gz:b-7p4az9) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/client.py#L171)):*
> ```Python
>     ADMIN_SERVICE = 'application/vnd.vmware.admin.service+xml'
>     API_EXTENSIBILITY = 'application/vnd.vmware.vcloud.apiextensibility+xml'
>     ADMIN_ORG = 'application/vnd.vmware.admin.organization+xml'
> ```


if we are moving constants around, let's arrange them in lexicographical order.

---

*[pyvcloud/vcd/client.py, line 844 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L1zYbouQ1S3GQd1a2OY:-L1zYbouQ1S3GQd1a2OZ:b-dalk) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/client.py#L844)):*
> ```Python
>         :param org_href: org where the user belongs.
> 
>         :return: User record
> ```

We say record but we are fetching references via the query. Please fix this.
viz. query_result_format=QueryResultFormat.REFERENCES,

ref : https://pubs.vmware.com/vca/index.jsp#com.vmware.vcloud.api.doc_56/GUID-4FD71B6D-6797-4B8E-B9F0-618F4ACBEFAC.html

---

*[pyvcloud/vcd/client.py, line 847 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L1zYHquChl9J7jkk5dC:-L1zYHquChl9J7jkk5dD:b-zeplce) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/client.py#L847)):*
> ```Python
> 
>         """
>         org_resource = self.get_resource(org_href)
> ```

If we are just interested in href of the org, why are we getting the whole resource using the href? We can directly use the href provided as a param.

---

*[pyvcloud/vcd/client.py, line 863 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L1zZVUXh5X_DF62uw9R:-L1zZVUXh5X_DF62uw9S:bzd78z9) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/client.py#L863)):*
> ```Python
>         elif len(records) > 1:
>             raise Exception('multiple users found')
>         return self.get_resource(records[0].get('href'))
> ```

This won't return a record but the complete lxml string representation of the resource!

---

*[tests/vcd_catalog.py, line 56 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L1za8NP_l6kbmq54iUm:-L1za8NQBiIpszGKtKZ4:b-rn7pk6) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/tests/vcd_catalog.py#L56)):*
> ```Python
>                  'access_level': 'ReadOnly'},
>                 {'name': self.config['vcd']['access_user'], 'type': 'user',
>                  'access_level':
> ```

I think we can fit the value and key both in the same line.

---

*[tests/vcd_catalog.py, line 119 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L1zafNvGdAB5jwPNBHU:-L1zafNvGdAB5jwPNBHV:b-fcxhy2) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/tests/vcd_catalog.py#L119)):*
> ```Python
>         assert control_access.IsSharedToEveryone.text == 'false'
> 
>     def test_10_remove_last_catalog_access(self):
> ```

What is the significance of this test?
why 'last' ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/163)*
<!-- Sent from Reviewable.io -->

pacogomez commented:
The logic behind this is: `vcd catalog acl list <catalog-name>` lists the `acl` entries of `<catalog-name>`. The `info` action provides all the `acl` related details of `<catalog-name>`, which includes the `is_shared_to_everyone` attribute, not shown in the `list` command. I agree that it departs from the semantics of `list` and `info` in other commands and we can revisit. The challenge is that `acl` items are not uniform, we have acl entries and is_shared_with_everyone attribute.

---

Review status: 3 of 4 files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-:-L20VntLTc-T7nob-nCk:b-heivvd)*
<!-- Sent from Reviewable.io -->

pacogomez commented:
On a second look, it seems like `share` is another way to set an ACL for the org where the catalog is defined. Is this a correct assumption? If yes, we could display all the ACL info in the the list, since is_shared_to_everyone is already represented by an ACL, in the example: `AnuOrg:org:None`

---

Review status: 3 of 4 files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-:-L20aYB3clnjl3k7ie-T:bckpbsu)*
<!-- Sent from Reviewable.io -->

sahithi commented:
Review status: 3 of 4 files reviewed at latest revision, 21 unresolved discussions.

---

*[pyvcloud/vcd/acl.py, line 32 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L2-0uDIHbE5LJJ-5ROM:-L22EXkt7x70XCAtYEwh:b32v57) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L32)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

I wouldn't call this resource, rather parent_resource. resource would be the XML we get when we do a GET on resource['href']/accesscontrol
</blockquote></details>

+1

---

*[pyvcloud/vcd/acl.py, line 35 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L2-2b2k066hIPpU_mU9:-L22ENvqBQTkRad9t48L:bu04bs0) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L35)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

Append a *new* access control list to the existing ACL of the parent resource
</blockquote></details>

+1
Append new access setting list to the existing ACL of the parent resource

---

*[pyvcloud/vcd/acl.py, line 81 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L22e8adDyRGnf81QSCH:-L22e8aetiTpuzin9CXH:b-2uzroz) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L81)):*
> ```Python
> 
>             # combine the new and old access settings
>             if hasattr(old_access_settings_params, 'AccessSetting'):
> ```

This check has to be done early (Check if the existing AccessSetting list is empty). It avoids running for loop search for common access_setting.

---

*[pyvcloud/vcd/acl.py, line 137 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L22cBnTRvwzBmbJ7oHm:-L22cBnTRvwzBmbJ7oHn:bqhimuu) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L137)):*
> ```Python
>                         old_access_settings_params)
>                 if matched_access_setting is None:
>                     raise Exception(
> ```

Print something like "Skipping to remove this particular Subject as it not found" and continue with removal of other subjects, instead of failing the entire operation.

---

*[pyvcloud/vcd/acl.py, line 144 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L22eibPmz7kR6eGkbHe:-L22eibPmz7kR6eGkbHf:b418hf) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L144)):*
> ```Python
>             # appending the the modified old_access_settings_params to
>             # control_access_params if at least 1 AccessSetting exist in it.
>             if hasattr(old_access_settings_params, 'AccessSetting'):
> ```

Same here, check this early in the method and re-adjust the logic accordingly.

---

*[pyvcloud/vcd/acl.py, line 230 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L22Z4l68qfjhrB3u9Io:-L22Z4l7rIEKaSczByPE:bfzqgnf) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L230)):*
> ```Python
>                 subject_name = access_setting['name']
>                 subject_type = EntityType.USER.value
>             else:
> ```

Please incorporate below logic to catch user errors. I think vcd-cli is the better place to catch user errors.

if (type ==user) {}
else if (type == org) {}
else raise Exception("Invalid subject type")

---

*[pyvcloud/vcd/acl.py, line 242 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L22_fxvt9jZDxNwCxtb:-L22_fxvt9jZDxNwCxtc:b-wtm10s) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L242)):*
> ```Python
>                           href=subject_href,
>                           type=subject_type),
>                 E.AccessLevel(access_setting['access_level'])
> ```

Ensure access_setting['access_level'] is not empty or None. If empty, set the default value to ReadOnly.

---

*[pyvcloud/vcd/org.py, line 677 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L22VooDqIaq_0a1rL0h:-L22VooEauTdIT_FqDhP:b1p2ntm) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/org.py#L677)):*
> ```Python
>         the updated access control setting of the catalog.
>         """
>         catalog_resource = self.get_catalog(name=catalog_name)
> ```

This is an unnecessary call if access_settings_list is empty or None. Please add the None or Emptylist check.

---

*[pyvcloud/vcd/org.py, line 695 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L22WL9szu6RWt4Ktjz-:-L22WL9tvufiunnUhwoW:bpknj4a) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/org.py#L695)):*
> ```Python
>             should be removed
> 
>         :return:  A :class:`lxml.objectify.StringElement` object representing
> ```

nit: Multiple colons

---

*[pyvcloud/vcd/org.py, line 697 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L22WW59_BD8MYMGBrw8:-L22WW59_BD8MYMGBrw9:b-rcw66m) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/org.py#L697)):*
> ```Python
>         :return:  A :class:`lxml.objectify.StringElement` object representing
>             the updated access control setting of the catalog.
>         """
> ```

Add Empty list or None check

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/163)*
<!-- Sent from Reviewable.io -->

sahithi commented:
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 21 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/163)*
<!-- Sent from Reviewable.io -->

anusuyar commented:
@pacogomez  Yes. is_shared_to_everyone=False can be displayed in the list command as Orgname:org:None and similarly is_shared_to_everyone=True can be displayed as OrgName:org:everyone_access_level ( I realized that I could do this during implementation).

I think in this case, I will remove the info command.

---

Review status: 1 of 4 files reviewed at latest revision, 22 unresolved discussions.

---

*[pyvcloud/vcd/acl.py, line 29 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L2-00kebISHl1KYO7t0:-L22Y8opSXWWR_ZhgRjS:b-896fix) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L29)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

nit : XML representation of the 'parent' entity whose Access Control List this object operates on.
</blockquote></details>

Done.

---

*[pyvcloud/vcd/acl.py, line 32 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L2-0uDIHbE5LJJ-5ROM:-L22Y8MDwNDRQUs1jNjP:b-896fix) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L32)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

I wouldn't call this resource, rather parent_resource. resource would be the XML we get when we do a GET on resource['href']/accesscontrol
</blockquote></details>

Done.

---

*[pyvcloud/vcd/acl.py, line 35 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L2-2b2k066hIPpU_mU9:-L22YATgMJ9YYZTwrIBc:b-896fix) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L35)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

Append a *new* access control list to the existing ACL of the parent resource
</blockquote></details>

Done.

---

*[pyvcloud/vcd/acl.py, line 53 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L2-3IHgU7T_GwXzriGn:-L22YCSlZonj8DVjvH3u:blpoltd) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L53)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

This check should be the first thing we do in this function, else we will be making a ReST call to vCD to get control_access_param for no reason.
Infact the control_access_param is the resource in this context! So if we cache that, we don't need to make the ReST call.
</blockquote></details>

Done. introduced a cache self.resource

---

*[pyvcloud/vcd/acl.py, line 60 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L2-351JKgDkqet6nixf:-L22YH0vZQBcwtkYaCSF:b-896fix) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L60)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

typo : fro ?
</blockquote></details>

Done.

---

*[pyvcloud/vcd/acl.py, line 66 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L2-5kbeiZgoeioTpNmR:-L22Y5GTEzMheNRVNUzd:b-pgj6vt) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L66)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

Is this conversion to list necessary? access_settings_list should be a list , right?
</blockquote></details>

Done. Removed the list()

---

*[pyvcloud/vcd/acl.py, line 81 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L22e8adDyRGnf81QSCH:-L23QeOUsckxpGlE74Dq:b-ur6uv2) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L81)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

This check has to be done early (Check if the existing AccessSetting list is empty). It avoids running for loop search for common access_setting.
</blockquote></details>

This check cannot be done early. 

First I am removing all matched_access_settings from old_access_settings_params
Then i am creating new_access_settings_params from access_settings_list

Then i am appending old_access_settings_params to new_access_settings_params  if old_access_settings_params has atleast 1 AssessSettings.  This check has to be here.

I am returning if  access_settings_list is None in the first line of the method in my next commit.

---

*[pyvcloud/vcd/acl.py, line 92 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L2-672rpHKg4keObceJ:-L22YIHiLJVQyFtXo5Up:b-896fix) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L92)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

We should change the if condition so that the cache-ed control_access_params is returned if access_settings_list if None, that will let us to indent the whole block of code above to left.
</blockquote></details>

Done.

---

*[pyvcloud/vcd/acl.py, line 137 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L22cBnTRvwzBmbJ7oHm:-L234feR1EE2QwxsXqbj:b-tav26c) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L137)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Print something like "Skipping to remove this particular Subject as it not found" and continue with removal of other subjects, instead of failing the entire operation.
</blockquote></details>

It was decided to throw an Exception when the subject was not found , instead of proceeding with the remaining subjects. I think it will be much simpler for the users.

---

*[pyvcloud/vcd/acl.py, line 144 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L22eibPmz7kR6eGkbHe:-L22xYdluuJSEgsJIOhI:b-27gcz) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L144)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Same here, check this early in the method and re-adjust the logic accordingly.
</blockquote></details>

This check is necessary here. We are removing matched_access_setting from  old_access_settings_params. We want to add this  old_access_settings_params to control_access_params only if there it at least 1 AssessSetting tag. If there is none, no need to append old_access_settings_params to control_access_params.  Modified the comment to make it more clear.

---

*[pyvcloud/vcd/acl.py, line 230 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L22Z4l68qfjhrB3u9Io:-L23Soexbc3IzePFuozd:b-896fix) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L230)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Please incorporate below logic to catch user errors. I think vcd-cli is the better place to catch user errors.

if (type ==user) {}
else if (type == org) {}
else raise Exception("Invalid subject type")
</blockquote></details>

Done.

---

*[pyvcloud/vcd/acl.py, line 242 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L22_fxvt9jZDxNwCxtb:-L23TPhArcgS8m8yWMFw:b-896fix) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L242)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Ensure access_setting['access_level'] is not empty or None. If empty, set the default value to ReadOnly.
</blockquote></details>

Done.

---

*[pyvcloud/vcd/client.py, line 171 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L1zXm0LVgdhf6kGP4gy:-L22YamMWL3xU7EgEFQy:b-896fix) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/client.py#L171)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

if we are moving constants around, let's arrange them in lexicographical order.
</blockquote></details>

Done.

---

*[pyvcloud/vcd/client.py, line 844 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L1zYbouQ1S3GQd1a2OY:-L23-y-PkAteu2wk7csM:b5iom8c) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/client.py#L844)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

We say record but we are fetching references via the query. Please fix this.
viz. query_result_format=QueryResultFormat.REFERENCES,

ref : https://pubs.vmware.com/vca/index.jsp#com.vmware.vcloud.api.doc_56/GUID-4FD71B6D-6797-4B8E-B9F0-618F4ACBEFAC.html
</blockquote></details>

Done. Removed the work record from the docstring

---

*[pyvcloud/vcd/client.py, line 847 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L1zYHquChl9J7jkk5dC:-L22wVYcoRV_ve3gJBIj:b-896fix) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/client.py#L847)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

If we are just interested in href of the org, why are we getting the whole resource using the href? We can directly use the href provided as a param.
</blockquote></details>

Done.

---

*[pyvcloud/vcd/client.py, line 863 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L1zZVUXh5X_DF62uw9R:-L2301rFUm78avD8jZBF:b-896fix) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/client.py#L863)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

This won't return a record but the complete lxml string representation of the resource!
</blockquote></details>

Done.

---

*[pyvcloud/vcd/org.py, line 677 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L22VooDqIaq_0a1rL0h:-L22uYZejJxTrDtfK5Ur:brhypd) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/org.py#L677)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

This is an unnecessary call if access_settings_list is empty or None. Please add the None or Emptylist check.
</blockquote></details>

Since we are returning the controlAccessParams for add_catalog_access_settings, the code here is necessary.  acl.add_access_settings is will return immediately. I think it is better for Acl class to deal with the logic instead of dealing with it here.   add_catalog_access_settings should only deal with calling the Acl methods. in the future if we want to change the logic, we can change   in one place in acl class methods.

---

*[pyvcloud/vcd/org.py, line 695 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L22WL9szu6RWt4Ktjz-:-L231MKkXDKmx1pO3F92:b-896fix) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/org.py#L695)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

nit: Multiple colons
</blockquote></details>

Done.

---

*[pyvcloud/vcd/org.py, line 697 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L22WW59_BD8MYMGBrw8:-L22x5_YkYewtpo2q1iH:biegono) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/org.py#L697)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Add Empty list or None check
</blockquote></details>

As above, the null check will be done in acl.remove_access_settings and that method will return immediately.

---

*[tests/vcd_catalog.py, line 56 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L1za8NP_l6kbmq54iUm:-L23WI1eRMyCzFFCzkwp:b-896fix) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/tests/vcd_catalog.py#L56)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

I think we can fit the value and key both in the same line.
</blockquote></details>

Done.

---

*[tests/vcd_catalog.py, line 119 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L1zafNvGdAB5jwPNBHU:-L234CLWHHCTlwZtBeVv:b-clw9wt) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/tests/vcd_catalog.py#L119)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

What is the significance of this test?
why 'last' ?
</blockquote></details>

I found a bug where removing the 1 remaining access_setting was causing errors. So added a TC  for that particular scenario.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-:-L21xjjZ2W18xRqr-Slt:bjp0f7h)*
<!-- Sent from Reviewable.io -->

rocknes commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Please run though others the design aspect of loading the resource inside constructor. Probably we should allow the user to provide just href of the parent resource and fetch parent_resource, resource etc. as and when needed.

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 18 unresolved discussions.

---

*[pyvcloud/vcd/acl.py, line 36 at r2](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L24LPWSIpBbmjDi7zqz:-L24LPWSIpBbmjDi7zr-:beoaord) ([raw file](https://github.com/vmware/pyvcloud/blob/5999d52c0f24b3239c043656b401450fc9f7c2fa/pyvcloud/vcd/acl.py#L36)):*
> ```Python
>         self.client = client
>         self.parent_resource = parent_resource
>         if resource is None:
> ```

Just fyi : This goes against the design we have in place viz. don't load things at construction time (lazy loading)

---

*[pyvcloud/vcd/acl.py, line 214 at r2](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L24Nojg_IMIRgHjfgJh:-L24Nojg_IMIRgHjfgJi:bc9u0s6) ([raw file](https://github.com/vmware/pyvcloud/blob/5999d52c0f24b3239c043656b401450fc9f7c2fa/pyvcloud/vcd/acl.py#L214)):*
> ```Python
>         # IsSharedToEveryone is false
>         if hasattr(control_access_params, 'EveryoneAccessLevel'):
>             everyone_access_level = getattr(control_access_params,
> ```

control_access_params.EveryoneAccessLevel should be sufficient to extract the child tag

---

*[pyvcloud/vcd/client.py, line 839 at r2](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L24OqN6flpD6tQLTdVL:-L24OqN6flpD6tQLTdVM:b8au3nd) ([raw file](https://github.com/vmware/pyvcloud/blob/5999d52c0f24b3239c043656b401450fc9f7c2fa/pyvcloud/vcd/client.py#L839)):*
> ```Python
> 
>     def get_user_in_org(self, user_name, org_href):
>         """Retrieve user  from a particular org.
> ```

nit: Stray space.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-:-L24PSSJMpAbN626k-JQ:b-d9881b)*
<!-- Sent from Reviewable.io -->

sompa commented:
Reviewed 1 of 4 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 18 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/163)*
<!-- Sent from Reviewable.io -->

pacogomez commented:
Review status: all files reviewed at latest revision, 18 unresolved discussions.

---

*[pyvcloud/vcd/acl.py, line 36 at r2](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L24LPWSIpBbmjDi7zqz:-L26fG-iK4-30bvMfTaU:b5ac3g5) ([raw file](https://github.com/vmware/pyvcloud/blob/5999d52c0f24b3239c043656b401450fc9f7c2fa/pyvcloud/vcd/acl.py#L36)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

Just fyi : This goes against the design we have in place viz. don't load things at construction time (lazy loading)
</blockquote></details>

that's right, we generally don't get the resource in the constructor, that's why we check if the resource is None on every method that might need the resource and load it on demand

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/163)*
<!-- Sent from Reviewable.io -->

anusuyar commented:
Review status: all files reviewed at latest revision, 18 unresolved discussions.

---

*[pyvcloud/vcd/acl.py, line 36 at r2](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L24LPWSIpBbmjDi7zqz:-L26nSdes_KHW4HN_6Do:b-a6o906) ([raw file](https://github.com/vmware/pyvcloud/blob/5999d52c0f24b3239c043656b401450fc9f7c2fa/pyvcloud/vcd/acl.py#L36)):*
<details><summary><i>Previously, pacogomez (Paco Gómez) wrote…</i></summary><blockquote>

that's right, we generally don't get the resource in the constructor, that's why we check if the resource is None on every method that might need the resource and load it on demand
</blockquote></details>

I will change the code to load the resource on demand.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/163)*
<!-- Sent from Reviewable.io -->

sahithi commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[pyvcloud/vcd/acl.py, line 81 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L22e8adDyRGnf81QSCH:-L26s8s0GeUARHztxlTq:b-pkdxdo) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L81)):*
<details><summary><i>Previously, anusuyar wrote…</i></summary><blockquote>

This check cannot be done early. 

First I am removing all matched_access_settings from old_access_settings_params
Then i am creating new_access_settings_params from access_settings_list

Then i am appending old_access_settings_params to new_access_settings_params  if old_access_settings_params has atleast 1 AssessSettings.  This check has to be here.

I am returning if  access_settings_list is None in the first line of the method in my next commit.


</blockquote></details>

Discussed offline with Anusuya, resolving the issue

---

*[pyvcloud/vcd/acl.py, line 137 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L22cBnTRvwzBmbJ7oHm:-L26i_55z9ZQzPqbMe0y:bqog8ju) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/acl.py#L137)):*
<details><summary><i>Previously, anusuyar wrote…</i></summary><blockquote>

It was decided to throw an Exception when the subject was not found , instead of proceeding with the remaining subjects. I think it will be much simpler for the users.
</blockquote></details>

Think of a use-case where command can have a list of 15 valid subjects and 1 invalid and entire cmd failing for the same. Think of a big script failing for user typo.

Anyway, I am unblocking this issue. Feel free to drop it if no one else has any issue.

---

*[pyvcloud/vcd/org.py, line 677 at r1](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L22VooDqIaq_0a1rL0h:-L26tTl5_oWHq8sshIsd:b-dhjbt9) ([raw file](https://github.com/vmware/pyvcloud/blob/1bfc14088d406d455a010729981a6e1f9f16d0bd/pyvcloud/vcd/org.py#L677)):*
<details><summary><i>Previously, anusuyar wrote…</i></summary><blockquote>

Since we are returning the controlAccessParams for add_catalog_access_settings, the code here is necessary.  acl.add_access_settings is will return immediately. I think it is better for Acl class to deal with the logic instead of dealing with it here.   add_catalog_access_settings should only deal with calling the Acl methods. in the future if we want to change the logic, we can change   in one place in acl class methods.
</blockquote></details>

After discussion with Anusuya, she pointed that vcd-cli has a check, so resolving the issue

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-:-L26tvmlLfKmDkEUs7PW:bnfp4nl)*
<!-- Sent from Reviewable.io -->

pacogomez commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 3 of 4 files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-:-L275gB0Hxv13v7nmsD_:bnfp4nl)*
<!-- Sent from Reviewable.io -->

anusuyar commented:
Review status: 2 of 4 files reviewed at latest revision, 11 unresolved discussions.

---

*[pyvcloud/vcd/acl.py, line 214 at r2](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L24Nojg_IMIRgHjfgJh:-L275RqJnlB6zQjIRpKN:b-896fix) ([raw file](https://github.com/vmware/pyvcloud/blob/5999d52c0f24b3239c043656b401450fc9f7c2fa/pyvcloud/vcd/acl.py#L214)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

control_access_params.EveryoneAccessLevel should be sufficient to extract the child tag
</blockquote></details>

Done.

---

*[pyvcloud/vcd/client.py, line 839 at r2](https://reviewable.io:443/reviews/vmware/pyvcloud/163#-L24OqN6flpD6tQLTdVL:-L275XGkisxz0Zhvq7gK:b-896fix) ([raw file](https://github.com/vmware/pyvcloud/blob/5999d52c0f24b3239c043656b401450fc9f7c2fa/pyvcloud/vcd/client.py#L839)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

nit: Stray space.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/163)*
<!-- Sent from Reviewable.io -->

pacogomez commented:
Reviewed 1 of 4 files at r1, 1 of 3 files at r2, 1 of 1 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vmware/pyvcloud/163)*
<!-- Sent from Reviewable.io -->
